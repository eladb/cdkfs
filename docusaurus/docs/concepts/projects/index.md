# Projects

Projects are the top level component in projen and the entry point for all other components.
They encompass many components, represent a specific type of codebase, and are opinionated, yet flexible.

For example, the [AwsCdkTypeScriptApp](/docs/api/awscdk#awscdktypescriptapp-) represents an [AWS CDK](https://aws.amazon.com/cdk/) application written in TypeScript. 
It includes opinionated defaults for the location of source files, the testing framework to use, and automation tools using GitHub Actions. 
While many of the specifics can be customized, the project type itself is opinionated and will always produce a codebase that is consistent with other projects of the same type.

There are many different types of projects that ship with projen, and you can also create your own.
Projects that ship with projen are grouped into namespaces, such as 'awscdk', 'java', and 'typescript'.
In the [Getting Started](/docs/getting-started), the [TypescriptAppProject](/docs/api/typescript#typescriptappproject-) was used, which comes from the `typescript` namespace.

```typescript
import { typescript } from 'projen';
const project = new typescript.TypeScriptAppProject({
  defaultReleaseBranch: 'main',
  name: 'my-project',
  projenrcTs: true,
  deps: ['express'], // added 'express' to the deps array
});
project.synth();
```

This produces a baseline Typescript application and adds the express dependency to the `package.json` file.
