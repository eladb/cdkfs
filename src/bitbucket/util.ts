

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}