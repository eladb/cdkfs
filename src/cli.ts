import * as fs from 'fs-extra';
import * as path from 'path';
import {PROJEN_RC} from './common';
import * as yargs from 'yargs';

const projenfile = path.resolve(PROJEN_RC);
const projen = path.join(__dirname, '..');

const args = yargs
  .commandDir('cmds')
  .recommendCommands()
  .wrap(yargs.terminalWidth())
  .help()
  .argv;

// no command means just require .projenrc.js
if (args._.length === 0) {
  if (!fs.existsSync(projenfile)) {
    console.error(`Unable to find ${projenfile}. Use "projen new" to create a new project.`);
    process.exit(1);
  }

  // now in order to run, projen needs to be able to "require('projen')". this will
  // work if we already installed our modules, but will fail miserably when we
  // bootstrap a checked out (or empty) repo.
  let unlink;
  const symlink = path.join('node_modules', 'projen');
  if (!fs.existsSync(symlink)) {
    fs.mkdirpSync('node_modules');
    fs.symlinkSync(projen, symlink);
    unlink = symlink;
  }

  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    require(projenfile);
  } finally {
    if (unlink) {
      fs.unlinkSync(symlink);
    }
  }
}
