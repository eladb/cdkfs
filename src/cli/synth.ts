import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs-extra';
import { PROJEN_RC } from '../common';
import * as logging from '../logging';

const projenfile = path.resolve(PROJEN_RC);
const projen = path.join(__dirname, '..');

export function synth() {
  if (!fs.existsSync(projenfile)) {
    logging.error(`Unable to find ${projenfile}. Use "projen new" to create a new project.`);
    process.exit(1);
  }

  // if node_modules/projen is not a directory or does not exist, create a
  // temporary symlink to the projen that we are currently running in order to
  // allow .projenrc.js to `require()` it.
  logging.info('Synthesizing project...');
  const projenModulePath = path.resolve('node_modules', 'projen');
  if (!fs.existsSync(path.join(projenModulePath, 'package.json')) || !fs.statSync(projenModulePath).isDirectory()) {
    fs.removeSync(projenModulePath);
    fs.mkdirpSync('node_modules');
    fs.symlinkSync(projen, projenModulePath, (os.platform() === 'win32') ? 'junction' : null);
  }

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  require(projenfile);
}