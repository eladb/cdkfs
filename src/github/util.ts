export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
