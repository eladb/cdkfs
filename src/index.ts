export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';