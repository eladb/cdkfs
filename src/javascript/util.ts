import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
