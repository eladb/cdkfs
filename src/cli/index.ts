import { resolve } from 'path';
import * as yargs from 'yargs';
import { PROJEN_RC } from '../common';
import { synth } from './synth';
import { discoverTaskCommands } from './tasks';

const DEFAULT_RC = resolve(PROJEN_RC);

async function main() {
  const ya = yargs;
  ya.commandDir('cmds');

  discoverTaskCommands(ya);

  ya.recommendCommands();
  ya.wrap(yargs.terminalWidth());
  ya.option('post', { type: 'boolean', default: true, desc: 'Run post-synthesis steps such as installing dependencies. Use --no-post to skip' });
  ya.options('debug', { type: 'boolean', default: false, desc: 'Debug logs' });
  ya.options('rc', { desc: 'path to .projenrc.js file', default: DEFAULT_RC, type: 'string' });
  ya.help();

  const args = ya.argv;

  if (args.debug) {
    process.env.DEBUG = 'true';
  }

  // no command means just require .projenrc.js
  if (args._.length === 0) {
    process.env.PROJEN_DISABLE_POST = (!args.post).toString();
    await synth((args.rc as string) ?? DEFAULT_RC);
  }
}


main().catch(e => {
  console.error(e.stack);
  process.exit(1);
});
