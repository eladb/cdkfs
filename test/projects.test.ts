import { installPackage } from '../src/cli/util';
import { NewProjectOptionHints } from '../src/option-hints';
import { Projects } from '../src/projects';
import { directorySnapshot, withProjectDir } from '../src/util/synth';

describe('createProject', () => {
  test('creates a project in a directory', () => {
    withProjectDir(projectdir => {
      // GIVEN
      Projects.createProject({
        optionHints: NewProjectOptionHints.FEATURED,
        dir: projectdir,
        post: false,
        synth: false,
        projectFqn: 'projen.typescript.TypeScriptProject',
        projectOptions: {
          name: 'test-project',
          defaultReleaseBranch: 'main',
        },
      });

      // THEN
      const snapshot = directorySnapshot(projectdir, {
        excludeGlobs: ['node_modules/**'],
      });
      expect(snapshot['.projenrc.js']).toMatchSnapshot();
    }, { chdir: true });
  });

  test('creates a project and passes in JSON-like project options', () => {
    withProjectDir(projectdir => {
      // GIVEN
      Projects.createProject({
        optionHints: NewProjectOptionHints.FEATURED,
        dir: projectdir,
        post: false,
        synth: false,
        projectFqn: 'projen.typescript.TypeScriptProject',
        projectOptions: {
          name: 'test-project',
          defaultReleaseBranch: 'main',
          eslintOptions: {
            dirs: ['src', 'test'],
            prettier: true,
            aliasMap: {
              '@src': './src',
              '@foo': './src/foo',
            },
          },
        },
      });

      // THEN
      const snapshot = directorySnapshot(projectdir, {
        excludeGlobs: ['node_modules/**'],
      });
      expect(snapshot['.projenrc.js']).toMatchSnapshot();
    }, { chdir: true });
  });

  test('creates a project from an external project type, if it\'s installed', () => {
    withProjectDir(projectdir => {
      // GIVEN
      installPackage(projectdir, '@taimos/projen@0.0.127');

      // WHEN
      Projects.createProject({
        optionHints: NewProjectOptionHints.FEATURED,
        dir: projectdir,
        post: false,
        synth: false,
        projectFqn: '@taimos/projen.TaimosTypescriptLibrary',
        projectOptions: {
          name: 'test-project',
          defaultReleaseBranch: 'main',
        },
      });

      // THEN
      const snapshot = directorySnapshot(projectdir, {
        excludeGlobs: ['node_modules/**'],
      });
      expect(snapshot['.projenrc.js']).toMatchSnapshot();
    }, { chdir: true });
  });
});
