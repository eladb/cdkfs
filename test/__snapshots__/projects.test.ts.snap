// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProject creates a project and passes in JSON-like project options 1`] = `
"const { TypeScriptProject } = require(\\"projen\\");
const project = new TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  eslintOptions: {\\"dirs\\":[\\"src\\",\\"test\\"],\\"prettier\\":true,\\"aliasMap\\":{\\"@src\\":\\"./src\\",\\"@foo\\":\\"./src/foo\\"}},
  name: \\"test-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();"
`;

exports[`createProject creates a project from an external project type, if it's installed 1`] = `
"const { AwsCdkAppSyncApp } = require('cdk-appsync-project');
const project = new AwsCdkAppSyncApp({
  cdkVersion: '1.63.0',
  defaultReleaseBranch: 'main',
  devDeps: ['cdk-appsync-project@1.1.3'],
  name: 'test-project',
  transformerVersion: '1.77.15',

  // cdkDependencies: undefined,        /* Which AWS CDK modules (those that start with \\"@aws-cdk/\\") this app uses. */
  // deps: [],                          /* Runtime dependencies of this module. */
  // description: undefined,            /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,            /* The \\"name\\" in package.json. */
  // projectType: ProjectType.UNKNOWN,  /* Which type of project this is (library/app). */
  // releaseWorkflow: undefined,        /* Define a GitHub workflow for releasing from \\"main\\" when new versions are bumped. */
});
project.synth();"
`;

exports[`createProject creates a project in a directory 1`] = `
"const { TypeScriptProject } = require(\\"projen\\");
const project = new TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"test-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();"
`;
